trinity/sync/beacon/chain.py

```diff
@@ -0,0 +1,145 @@
+from typing import (
+    cast,
+    Tuple,
+    AsyncGenerator,
+)
+
+from eth_utils import (
+    ValidationError,
+)
+from eth_utils.toolz import (
+    first,
+)
+
+from cancel_token import (
+    CancelToken,
+)
+
+from p2p.service import (
+    BaseService,
+)
+
+from eth.beacon.types.blocks import BaseBeaconBlock
+from eth.beacon.db.chain import BaseBeaconChainDB
+
+from trinity.protocol.bcc.peer import (
+    BCCPeer,
+    BCCPeerPool,
+)
+from trinity.sync.beacon.constants import (
+    MAX_BLOCKS_PER_REQUEST,
+)
+
+
+class BeaconChainSyncer(BaseService):
+    """Sync from our finalized head until their preliminary head."""
+
+    def __init__(self,
+                 chain_db: BaseBeaconChainDB,
+                 peer_pool: BCCPeerPool,
+                 token: CancelToken = None) -> None:
+        super().__init__(token)
+
+        self.chain_db = chain_db
+        self.peer_pool = peer_pool
+
+        self.sync_peer: BCCPeer = None
+
+    async def _run(self) -> None:
+        self.sync_peer = self.select_sync_peer()
+        if not self.sync_peer:
+            self.logger.info("No peers to sync with")
+            return
+
+        if not self.sync_peer_useful():
+            self.logger.info("Not starting sync as no peer is useful")
+            return
+
+        await self.sync()
+
+        new_head = self.chain_db.get_canonical_head()
+        self.logger.info(f"Sync with {self.sync_peer} finished, new head: {new_head}")
+
+    def select_sync_peer(self) -> BCCPeer:
+        if len(self.peer_pool) == 0:
+            raise ValueError("No peers to sync with")
+        peers = [
+            cast(BCCPeer, peer)
+            for peer in self.peer_pool.connected_nodes.values()
+        ]
+
+        # chooses the peer with the highest head slot
+        sorted_peers = sorted(peers, key=lambda peer: peer.head_slot, reverse=True)
+        best_peer = first(sorted_peers)
+        return best_peer
+
+    def sync_peer_useful(self) -> bool:
+        return self.chain_db.get_finalized_head().slot < self.sync_peer.head_slot
+
+    async def sync(self) -> None:
+        self.logger.info(
+            "Syncing with %s (their head slot: %d, our finalized slot: %d)",
+            self.sync_peer,
+            self.sync_peer.head_slot,
+            self.chain_db.get_finalized_head().slot,
+        )
+
+        start_slot = self.chain_db.get_finalized_head().slot + 1
+        batches = self.request_batches(start_slot)
+
+        last_block = None
+        async for batch in batches:
+            if last_block is None:
```

Comment:
This seems like a hack to do something special on the *first* item.  Maybe change to something like:

```python
async def async_enumerate(iterable, start_idx=None):
    counter = itertools.counter(start_idx)
    async for value in iterable:
        yield next(counter), value

    async for (idx, batch) in async_enumerate(batches):
        if idx == 0:
            ....
```

A more ideal solution ignoring the async stuff would be:

```python
for previous_batch, batch in sliding_window(2, concatv([None], batches)):
    if previous_batch is None:
        self.validate_first_batch(batch)
    elif batch[0].parent_root != previous_batch[-1].hash:
        ...
```

But the async nature of `batches` complicates this approach.  I suspect we could generate some basic machinery to handle that though and I think the resulting body of this `for` loop could be cleaner.  I'll leave it to you to decide what (if anything) you'd like to do here.